src/vs/base/common/glob.ts

```diff
@@ -817,3 +817,108 @@ export function patternsEquals(patternsA: Array<string | IRelativePattern> | und
 		return false;
 	});
 }
+
+// brace expansion
+
+/**
+ * Split string given first opportunity for brace expansion in the string.
+ * - If the brace is prepended by a \ character, then it is escaped.
+ * - Does not process escapes that are within the sub-glob.
+ * - If two unescaped `{` occur before `}`, then ripgrep will return an error for brace nesting, so don't split on those.
+ */
+function getEscapeAwareSplitString(pattern: string): { fixedStart?: string; strInBraces: string; fixedEnd?: string } {
+	let inBraces = false;
+	let escaped = false;
+	let fixedStart = '';
+	let strInBraces = '';
+	for (let i = 0; i < pattern.length; i++) {
+		const char = pattern[i];
+		switch (char) {
+			case '\\':
+				if (escaped) {
+					// If we're already escaped, then just leave the escaped slash and the preceeding slash that escapes it.
+					// The two escaped slashes will result in a single slash and whatever processes the glob later will properly process the escape
+					if (inBraces) {
+						strInBraces += '\\' + char;
+					} else {
+						fixedStart += '\\' + char;
+					}
+					escaped = false;
+				} else {
+					escaped = true;
+				}
+				break;
+			case '{':
+				if (escaped) {
+					// if we escaped this opening bracket, then it is to be taken literally. Remove the `\` because we've acknowleged it and add the `{` to the appropriate string
+					if (inBraces) {
+						strInBraces += char;
+					} else {
+						fixedStart += char;
+					}
+					escaped = false;
+				} else {
+					if (inBraces) {
+						// ripgrep treats this as attempting to do an alternating group, which is invalid. Return with pattern including changes from escaped braces.
+						return { strInBraces: fixedStart + '{' + strInBraces };
+					} else {
+						inBraces = true;
+					}
+				}
+				break;
+			case '}':
+				if (escaped) {
+					// same as `}`, but for closing bracket
+					if (inBraces) {
+						strInBraces += char;
+					} else {
+						fixedStart += char;
+					}
+					escaped = false;
+				} else if (inBraces) {
+					// we found an end bracket to a valid opening bracket. Return the appropriate strings.
+					return { fixedStart, strInBraces, fixedEnd: pattern.substring(i + 1) };
```

Comment:
Does this mean you can only have one `{}` pattern per glob? Is that the rule in glob.ts in general or in ripgrep?

Reply:
this function finds the first valid `{` and `}` pair, but the caller calls this function again on the trailing part of the string to parse the rest of it